### PR TITLE
retrofit the scheduler with the leader election client.

### PR DESCRIPTION
**Pull request by mikedanese**
cc @timothysc

### CODE REVIEW

File: plugin/cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -110,6 +122,11 @@ func (s *SchedulerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver")
 	fs.IntVar(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.StringVar(&s.SchedulerName, "scheduler-name", s.SchedulerName, "Name of the scheduler, used to select which pods will be processed by this scheduler, based on pod's annotation with key 'scheduler.alpha.kubernetes.io/name'")
+
+	fs.BoolVar(&s.LeaderElect, "leader-elect", s.LeaderElect, "Start a leader election client and gain leadership before executing scheduler loop.")
```
Comment (lavalamp) [inline plugin/cmd/kube-scheduler/app/server.go]: A driveby thought which may or may not be a good idea: If this is going to be a common pattern, consider putting the code for setting up these flags into the leaderelection package?

Comment (mikedanese) [inline plugin/cmd/kube-scheduler/app/server.go]: @lavalamp good idea, done.

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by mikedanese** [comment]
@davidopp I've attempted to clarify wording on flag description. Can you please comment on what you think is still vague?

File: docs/admin/kube-scheduler.md [hunk 2]
```diff
@@ -62,6 +62,10 @@ kube-scheduler
       --kube-api-burst=100: Burst to use while talking with kubernetes apiserver
       --kube-api-qps=50: QPS to use while talking with kubernetes apiserver
       --kubeconfig="": Path to kubeconfig file with authorization and master location information.
+      --leader-elect[=false]: Start a leader election client and gain leadership before executing scheduler loop. Enable this when running replicated schedulers.
+      --leader-elect-lease-duration=15s: Lease duration is the duration that non-leader candidates will wait until attempting to acquire leadership of a led but unrenewed leader slot. This is effectively the maximum duration that a leader can be stopped before it is replaced by another candidate. This is only applicable if leader election is enabled.
```
Comment (davidopp) [inline docs/admin/kube-scheduler.md]: s/Lease duration is the//

rewrite is much better -- only part people might find confusing is "will wait" -- will wait after what? "led but unrenewed" gives a good hint but maybe clearer to say "will wait after the most recent renewal" or something like that?

File: docs/admin/kube-scheduler.md [hunk 3]
```diff
@@ -62,6 +62,10 @@ kube-scheduler
...
+      --leader-elect-renew-deadline=10s: Renew deadline is the duration that the acting master will retry renewing a leadership slot before it stops leadeing. This must be less than or equal to the lease duration. This is only applicable if leader election is enabled.
```
Comment (davidopp) [inline docs/admin/kube-scheduler.md]: s/Renew deadline is the//
s/duration that the acting master will retry renewing/interval between attempts by the acting master to renew/
also I'm not sure you should say "before it stops leading" -- IIUC it stops leading because someone else takes over, not because it does something?

File: docs/admin/kube-scheduler.md [hunk 4]
```diff
@@ -62,6 +62,10 @@ kube-scheduler
       --kube-api-burst=100: Burst to use while talking with kubernetes apiserver
       --kube-api-qps=50: QPS to use while talking with kubernetes apiserver
       --kubeconfig="": Path to kubeconfig file with authorization and master location information.
+      --leader-elect[=false]: Start a leader election client and gain leadership before executing scheduler loop. Enable this when running replicated schedulers.
```
Comment (timothysc) [inline docs/admin/kube-scheduler.md]: Is there any issue in O(1) case to not default to on?

File: docs/admin/kube-scheduler.md (hunk 3 above)
Comment (timothysc) [inline docs/admin/kube-scheduler.md]: s/leadeing/leading

File: docs/admin/kube-scheduler.md [hunk 5]
```diff
@@ -62,6 +62,10 @@ kube-scheduler
...
+      --leader-elect-lease-duration=15s: Lease duration is the duration that non-leader candidates will wait until attempting to acquire leadership of a led but unrenewed leader slot. This is effectively the maximum duration that a leader can be stopped before it is replaced by another candidate. This is only applicable if leader election is enabled.
+      --leader-elect-renew-deadline=10s: Renew deadline is the duration that the acting master will retry renewing a leadership slot before it stops leadeing. This must be less than or equal to the lease duration. This is only applicable if leader election is enabled.
+      --leader-elect-retry-period=2s: Retry period is the duration the LeaderElector clients should wait between actions against the API. Common actions are attempting to aquire the leader slot, renewing  leadership of a leader slot, or observing any updates of a leader slot. This is only applicable if leader election is true.
```
Comment (timothysc) [inline docs/admin/kube-scheduler.md]: this sentence could probably be trimmed : 
Retry period is the duration the clients should wait between lease acquisition and renewals.  This is only applicable if leader election is true.

File: pkg/client/leaderelection/leaderelection.go [hunk 6]
```diff
@@ -296,3 +308,50 @@ func (le *LeaderElector) tryAcquireOrRenew() bool {
 	le.observedTime = time.Now()
 	return true
 }
+
+func DefaultLeaderElectionCLIConfig() LeaderElectionCLIConfig {
+	return LeaderElectionCLIConfig{
+		LeaderElect:   false,
+		LeaseDuration: 15 * time.Second,
```
Comment (timothysc) [inline pkg/client/leaderelection/leaderelection.go]: given that these are consts... 
could we place the consts at the top, at least for consistency with other libs as they are cleaned.

File: plugin/cmd/kube-scheduler/app/options/options.go [hunk 7]
```diff
@@ -72,4 +75,5 @@ func (s *SchedulerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver")
 	fs.IntVar(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.StringVar(&s.SchedulerName, "scheduler-name", s.SchedulerName, "Name of the scheduler, used to select which pods will be processed by this scheduler, based on pod's annotation with key 'scheduler.alpha.kubernetes.io/name'")
+	s.LeaderElection.BindFlags(fs)
```
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/options/options.go]: We should be doing this **everywhere**!

**Comment by timothysc** [comment]
Seems like we should have a simple integration or e2e for this.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 8]
```diff
@@ -110,9 +111,41 @@ func Run(s *options.SchedulerServer) error {
...
 	sched := scheduler.New(config)
-	sched.Run()
 
-	select {}
+	run := func(_ <-chan struct{}) {
+		sched.Run()
+		select {}
+	}
+
+	if !s.LeaderElection.LeaderElect {
```
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/server.go]: Same comment re: is there any reason to not have this always enabled?

File: docs/admin/kube-scheduler.md (hunk 4 above)
Comment (mikedanese) [inline docs/admin/kube-scheduler.md]: I'm okay with that but I'd like to enable it by default only once we have an e2e test.

**Comment by mikedanese** [comment]
@timothysc I'm going to do controller-manager sperately as well as add an e2e test before v1.2 (so within the next week or two).

File: pkg/client/leaderelection/leaderelection.go [hunk 9]
```diff
@@ -296,3 +312,49 @@ func (le *LeaderElector) tryAcquireOrRenew() bool {
...
+type LeaderElectionCLIConfig struct {
+	LeaderElect   bool
+	LeaseDuration time.Duration
+	RenewDeadline time.Duration
+	RetryPeriod   time.Duration
+}
+
+// BindFlags binds the common LeaderElectionCLIConfig flags to a flagset
+func (l *LeaderElectionCLIConfig) BindFlags(fs *pflag.FlagSet) {
+	clean := func(s string) string {
```
Comment (davidopp) [inline pkg/client/leaderelection/leaderelection.go]: What is the purpose of this? Is it just so you can split each string across multiple lines? Why not just put it in one (very long) line?

File: docs/admin/kube-scheduler.md (hunk 4 above)
Comment (davidopp) [inline docs/admin/kube-scheduler.md]: Maybe say "replicated scheduler for high availability" just to make sure people don't confuse this with running multiple schedulers for performance or partitioning diffrent scheduling algorithms.

File: pkg/client/leaderelection/leaderelection.go (hunk 9 above)
Comment (mikedanese) [inline pkg/client/leaderelection/leaderelection.go]: Ya I couldn't get gofmt to do anything sane with multiline comments. I'll put it on one line.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 10]
```diff
@@ -110,9 +111,41 @@ func Run(s *options.SchedulerServer) error {
...
+		run(nil)
+		return fmt.Errorf("this is unreachable")
```
Comment (davidopp) [inline plugin/cmd/kube-scheduler/app/server.go]: probably a log(error) would be better

File: plugin/cmd/kube-scheduler/app/server.go [hunk 11]
```diff
@@ -110,9 +111,41 @@ func Run(s *options.SchedulerServer) error {
...
+		RenewDeadline: s.LeaderElection.RenewDeadline,
+		RetryPeriod:   s.LeaderElection.RetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Fatalf("lost master")
+			},
+		},
+	})
+	return fmt.Errorf("this is unreachable")
```
Comment (davidopp) [inline plugin/cmd/kube-scheduler/app/server.go]: probably a log(error) would be better

File: plugin/cmd/kube-scheduler/app/server.go [hunk 12]
```diff
@@ -110,9 +111,41 @@ func Run(s *options.SchedulerServer) error {
...
-	sched.Run()
 
-	select {}
+	run := func(_ <-chan struct{}) {
+		sched.Run()
+		select {}
+	}
+
+	if !s.LeaderElection.LeaderElect {
+		run(nil)
```
Comment (davidopp) [inline plugin/cmd/kube-scheduler/app/server.go]: I must be missing something -- while can't you just make this block say
```
sched.Run()
select {}
```
instead of having a separate run() function? Is run() used from somewhere else that I missed?

Comment (mikedanese) [inline plugin/cmd/kube-scheduler/app/server.go]: It's used again on line 43

Comment (davidopp) [inline plugin/cmd/kube-scheduler/app/server.go]: line 43 is an import line AFAICT

Comment (mikedanese) [inline plugin/cmd/kube-scheduler/app/server.go]: @davidopp sorry 143

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by davidopp** [comment]
LGTM

**Comment by mikedanese** [comment]
@k8s-bot unit test this please

File: docs/admin/kube-scheduler.md (hunk 4 above)
Comment (timothysc) [inline docs/admin/kube-scheduler.md]: @davidopp - Good point, defaulting on would be bad in this case.

**Comment by timothysc** [comment]
So we still need an e2e or integration on this.
